File: Stripe/STPPaymentMethodIdeal.m
//
//  STPPaymentMethodiDEAL.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/9/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPPaymentMethodiDEAL.h"

#import "NSDictionary+Stripe.h"

@interface STPPaymentMethodiDEAL ()

@property (nonatomic, nullable, copy, readwrite) NSString *bankName;
@property (nonatomic, nullable, copy, readwrite) NSString *bankIdentifierCode;
@property (nonatomic, copy, nonnull, readwrite) NSDictionary *allResponseFields;

@end

@implementation STPPaymentMethodiDEAL

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                       // Properties
                       [NSString stringWithFormat:@"bank: %@", self.bankName],
                       [NSString stringWithFormat:@"bic: %@", self.bankIdentifierCode],
                       ];

    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPAPIResponseDecodable

+ (instancetype)decodedObjectFromAPIResponse:(NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    STPPaymentMethodiDEAL *ideal = [self new];
    ideal.allResponseFields = dict;
    ideal.bankName = [dict stp_stringForKey:@"bank"];
    ideal.bankIdentifierCode = [dict stp_stringForKey:@"bic"];
    return ideal;
}

@end

File: Stripe/STPPaymentMethodiDEAL.swift
//
//  STPPaymentMethodiDEAL.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/9/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation

/// An iDEAL Payment Method.
/// - seealso: https://stripe.com/docs/api/payment_methods/object#payment_method_object-ideal
public class STPPaymentMethodiDEAL: NSObject, STPAPIResponseDecodable {
  /// The customer’s bank.
  @objc public private(set) var bankName: String?
  /// The Bank Identifier Code of the customer’s bank.
  @objc public private(set) var bankIdentifierCode: String?
  @objc public private(set) var allResponseFields: [AnyHashable: Any] = [:]

  /// :nodoc:
  @objc public override var description: String {
    let props = [
      // Object
      String(format: "%@: %p", NSStringFromClass(STPPaymentMethodiDEAL.self), self),
      // Properties
      "bank: \(bankName ?? "")",
      "bic: \(bankIdentifierCode ?? "")",
    ]

    return "<\(props.joined(separator: "; "))>"
  }

  // MARK: - STPAPIResponseDecodable
  override required init() {
    super.init()
  }

  @objc
  public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
    guard let response = response else {
      return nil
    }
    let dict = (response as NSDictionary).stp_dictionaryByRemovingNulls() as NSDictionary

    let ideal = self.init()
    ideal.allResponseFields = response
    ideal.bankName = dict.stp_string(forKey: "bank")
    ideal.bankIdentifierCode = dict.stp_string(forKey: "bic")
    return ideal
  }
}
